fix(aio): set the pageId to the file-not-found URL if the doc is not available

petebacondarwin opened the pull request:
Previously, the `AppComponent.pageId` was set via the current URL, rather than
the document being displayed. This is only really noticeable when the URL does not
match a valid doc and we are actually displaying a 404 page.

Now we compute the `pageId` from the  URL of the document being viewed,
which is returned from the  `DocumentService.currentDocument` observable instead.

petebacondarwin commented:
Oops I forgot to add the test of the fix 😱

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -66,14 +66,21 @@ export class AppComponent implements OnInit {
-    this.documentService.currentDocument.subscribe(doc => this.currentDocument = doc);
+    this.documentService.currentDocument.subscribe(doc => {
+      this.currentDocument = doc;
+      this.pageId = this.currentDocument && this.currentDocument.url.replace('/', '-');
```
_this.currentDocument_ --> _doc_ (?)

Can `doc` really be falsy? Looking at the `DocumentService`, I think not.
Also, `doc` might not have a `url` property (but it might actually be a `DocumentService` bug).

petebacondarwin replied:
I agree. I think it was from a dodgy test, which I have fixed

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -66,14 +66,21 @@ export class AppComponent implements OnInit {
+      this.pageId = this.currentDocument && this.currentDocument.url.replace('/', '-');
+
+      // Special case the home page
+      if (this.pageId === 'index') {
```
Why is the logic here different than before? Was it a bug in the previous implementation (where `pageId` was set to `home` only if falsy)?

petebacondarwin replied:
Previously we were getting the url from the navigationService, which returns `""` for the home page.
Now we are getting the url from the documentService, which returns `"index"` for the home page.

gkalpak commented on aio/src/app/documents/document.service.ts:
```diff
@@ -34,27 +33,29 @@ export class DocumentService {
+            return of({ title: 'Not Found', contents: 'Document not found', url: FILE_NOT_FOUND_URL });
           }
         } else {
           this.logger.error('Error fetching document', error);
```
I can't comment on the next line, but do we need to set a `url` there too?

petebacondarwin replied:
yes you are right

mary-poppins commented:
The angular.io preview for 7c1a895455b47769c1b6f1e4cd360548a00a2c19 is available [here][1].

[1]: https://pr16085-7c1a895455b47769c1b6f1e4cd360548a00a2c19.ngbuilds.io/

petebacondarwin committed:
fix(aio): set the pageId to the file-not-found URL if the doc is not available

Previously, the `AppComponent.pageId` was set via the current URL, rather than
the document being displayed. This is only really noticeable when the URL does not
match a valid doc and we are actually displaying a 404 page.

Now we compute the `pageId` from the  URL of the document being viewed,
which is returned from the  `DocumentService.currentDocument` observable instead.

petebacondarwin commented:
@gkalpak PTAL

mary-poppins commented:
The angular.io preview for 49c0cd0c7b33a6d94283167575af15906b734173 is available [here][1].

[1]: https://pr16085-49c0cd0c7b33a6d94283167575af15906b734173.ngbuilds.io/

mary-poppins commented:
The angular.io preview for b7b94a8c49b7becb2404917ee439635f11414d56 is available [here][1].

[1]: https://pr16085-b7b94a8c49b7becb2404917ee439635f11414d56.ngbuilds.io/

mary-poppins commented:
The angular.io preview for f3e681217795347ea295a14cd9330123d588cc79 is available [here][1].

[1]: https://pr16085-f3e681217795347ea295a14cd9330123d588cc79.ngbuilds.io/
